Add `ModalVector`

moxcodes opened the pull request:
## Proposed changes

This is based on the `ModalVector` changes implemented by @prayush, but updated
to fit into the framework with a templated `VectorImpl`. Multiplication,
division, and absolute values are omitted for `ModalVector`s, as these
operations do not have a clear and present use-case for spectral coefficients.

closes #378 

### Types of changes:

- [ ] New feature

### Component:

- [ ] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

## Dependencies
- [x] #970

moxcodes committed:
Add ModalVector

- A vector class intended for the storage of spectral coefficient data
- constructed with the `VectorImpl` framework
- Many mathematical operations are forbidden for `ModalVector`s, as these
operations do not have a clear use-case for spectral coefficients.

nilsdeppe commented on src/DataStructures/ModalVector.hpp:
```diff
@@ -0,0 +1,88 @@
+};
+BLAZE_TRAIT_SPEC_BINTRAIT(ModalVector, AddTrait);
+BLAZE_TRAIT_SPEC_BINTRAIT(ModalVector, SubTrait);
+BLAZE_TRAIT_COMPATIBLE_BINTRAIT(ModalVector, DataVector, AddTrait);
```
I think we don't want to be able to mix any operations between `ModalVector` and `DataVector`, right @kidder ?

kidder replied:
No, I don't know what that even means

nilsdeppe commented on tests/Unit/DataStructures/Test_ModalVector.cpp:
```diff
@@ -0,0 +1,99 @@
+
+class DataVector;
+
+SPECTRE_TEST_CASE("Unit.DataStructures.ModalVector", "[DataStructures][Unit]") {
```
Could you (similarly to DataVector now) combine the `ModalVector` and `ModalVector.Math` tests?

nilsdeppe commented on tests/Unit/DataStructures/Test_ModalVector.cpp:
```diff
@@ -0,0 +1,99 @@
+class DataVector;
+
+SPECTRE_TEST_CASE("Unit.DataStructures.ModalVector", "[DataStructures][Unit]") {
+  SECTION("test construct and assign") {
```
We'll want to follow the `INFO` things you've started doing :)

nilsdeppe commented on tests/Unit/DataStructures/Test_ModalVector.cpp:
```diff
@@ -0,0 +1,99 @@
+}
+
+// [[OutputRegex, Must copy into same size]]
+[[noreturn]] SPECTRE_TEST_CASE(
```
I think there's a third error test that does it for expressions of different size.

nilsdeppe commented on src/DataStructures/ModalVector.hpp:
```diff
@@ -0,0 +1,88 @@
+ * or non-owning, meaning it just has a pointer to an array.
+ *
+ * Only basic mathematical operations are supported with ModalVectors. In
+ * addition to addition, subtraction, multiplication, division, there
```
This sentence makes it sound like multiplying two `ModalVector`s is allowed, which I don't think it is (or should be). A different way to do the dox would be to write:

```cpp
The following mathematical binary expressions are allowed:
- ModalVector-ModalVector subtraction
- ModalVector-ModalVector addition
- double-ModalVector multiplication
- double-ModalVector division
- ModalVector-ModalVector max
- ModalVector-ModalVector min
```

I'm not sure how you feel about that but it might be a bit more explicit.

nilsdeppe reviewed (commented):
I realize we're not quite ready for this yet, but since it's really simple given all the other hard work you've done I figured I'd give feedback now. Nothing major :)

moxcodes committed:
Add DiagonalModalOperator

- used for element-wise multiplication of a set of data with a ModalVector, such
  as a filter or a set of derivative factors.

moxcodes commented:
This has been updated. I think I've addressed all of the review comments. In addition, I've added a class `DiagonalModalOperator` which is intended to be used as the container with which element-wise multiplication of modal coefficients is supported, for applications like filtering and derivatives.
Cheers!

moxcodes commented:
pushed again to fix include bug and to squash the dependency commits into one.

nilsdeppe commented on src/DataStructures/ModalVector.hpp:
```diff
@@ -0,0 +1,108 @@
+
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief A class for storing spectral coefficients on a mesh.
```
maybe `mesh` -> `spectral mesh` or `spectral grid` to emphasize that everything is in spectral space? It might be a bit redundant but I could see that being useful for someone newer to spectral-type methods.

nilsdeppe commented on src/DataStructures/ModalVector.hpp:
```diff
@@ -0,0 +1,108 @@
+ * ModalVectors. ModalVectors may be added or subtracted, and the following
+ * unary operations are supported:
+ * - max
+ * - min
```
Do we want to support `abs`? For AMR operations it is quite common to look at the absolute value of the coefficients.

moxcodes replied:
from the old conversations on #704, @kidder was questioning whether the absolute value of a `ModalVector` should be another `ModalVector`. Though, perhaps the practicality of wanting to numerically compute several absolute values quickly should rule in favor of permitting the operation.

I'll go ahead and add it; I just want to confirm that all of the type checking is as we intended.

nilsdeppe replied:
I talked with @kidder in person and we agreed that the result should be a `ModalVector` :)

nilsdeppe commented on src/DataStructures/Tensor/Tensor.hpp:
```diff
@@ -81,9 +82,10 @@ class Tensor<X, Symm, IndexList<Indices...>> {
           cpp17::is_same_v<X, std::complex<double>> or
           cpp17::is_same_v<X, ComplexDataVector>,
-      "Only a Tensor<double>, Tensor<DataVector>, "
+      "Only a Tensor<double>, Tensor<DataVector>, Tensor<ModalVector>"
```
you're missing a `, ` (comma and a space) at the end here

nilsdeppe commented on src/DataStructures/DiagonalModalOperator.hpp:
```diff
@@ -0,0 +1,108 @@
+ * the `DiagonalModalOperator` goes out of scope) or non-owning, meaning it just
+ * has a pointer to an array.
+ *
+ * Only basic mathematical operations are supported with
```
It might be good to state that this is essentially a diagonal matrix that can be applied to the spectral coefficients efficiently. While that can be deduced from everything else you say I think being explicit would be good :)

nilsdeppe reviewed (commented):
I'm happy for you to squash things immediately :)

moxcodes commented:
Changes squashed in.
Cheers!

nilsdeppe reviewed (dismissed):
@kidder @wthrowe could you please take a look at this when you get a chance? Thank you!

wthrowe commented on src/DataStructures/ModalVector.hpp:
```diff
@@ -0,0 +1,109 @@
+ * - min
+ *
+ * Also, addition, subtraction, multiplication, and division are supported with
+ * doubles and `ModalVector`s and multiplication is supported with
```
Does double + ModalVector make sense?

kidder replied:
No, it can't, same with subtraction.

kidder replied:
Also it should be clarified that only ModalVector/double is supported

wthrowe reviewed (commented):
I may not get a chance to re-review until Sunday or Monday, but you all can merge without waiting if you want to.

kidder commented on src/DataStructures/ModalVector.hpp:
```diff
@@ -0,0 +1,109 @@
+ * ModalVectors. ModalVectors may be added or subtracted, and the following
+ * unary operations are supported:
+ * - abs
+ * - max
```
I don't understand how `max` or `min` make any sense.

kidder commented on src/DataStructures/ModalVector.hpp:
```diff
@@ -0,0 +1,109 @@
+BLAZE_TRAIT_SPECIALIZE_BINARY_TRAIT(ModalVector, AddTrait);
+BLAZE_TRAIT_SPECIALIZE_BINARY_TRAIT(ModalVector, SubTrait);
+BLAZE_TRAIT_SPECIALIZE_COMPATIBLE_BINARY_TRAIT(ModalVector, double, MultTrait);
+BLAZE_TRAIT_SPECIALIZE_COMPATIBLE_BINARY_TRAIT(ModalVector, double, DivTrait);
```
does this only allow division by double?

moxcodes replied:
The macro permits either ordering, so I'll revise it to specialize only the appropriate operation (`ModalVector` divided by `double`s).

moxcodes committed:
Add option GivenOrderOfArgumentsOnly to vector math tests

- option indicates that only the specified combination of arguments is to be
  tested

moxcodes commented:
Fixups posted.
Cheers!

kidder commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -401,7 +401,12 @@ using Bound = std::array<double, 2>;
-enum TestKind { Normal, Strict, Inplace };
+///
+/// - `NoArgumentCombinations` is used to indicate that the arguments given
+///   should not be taken combinations apart from the given combination. This
```
in any combination apart from

kidder commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -401,7 +401,12 @@ using Bound = std::array<double, 2>;
 ///   `double += complex<double>` does not)
-enum TestKind { Normal, Strict, Inplace };
+///
+/// - `NoArgumentCombinations` is used to indicate that the arguments given
```
How about NoArgumentCombinations -> GivenOrderOfArgumentsOnly

nilsdeppe commented:
@kidder could you please check the fixups? Thanks! :)

nilsdeppe commented on src/DataStructures/DiagonalModalOperator.hpp:
```diff
@@ -0,0 +1,110 @@
+#include "Utilities/PointerVector.hpp"  // IWYU pragma: keep
+#include "Utilities/TMPL.hpp"
+
+class ModalVector;
```
`/// \cond` here

kidder reviewed (commented):
I am happy for you to squash

nilsdeppe reviewed (dismissed):
One minor thing I noticed while comparing to the complex modal vector :) I'm happy for squash time :)

moxcodes commented:
Squashed.
Cheers!

moxcodes commented:
There is a clang-tidy error, but it is mysterious to me. I am unable to replicate the error locally, and the error appears to refer to a state of the `VectorImplTestHelpers.hpp` that is different from the one provided by this PR.

kidder commented:
try rebasing on develop

moxcodes commented:
Rebased, all the clang-tidy checks pass now. Travis is failing only on gcc timeouts.
Cheers!

kidder commented:
travis failures are timeouts
